pyro/infer/mcmc/adaptation.py

```diff
@@ -350,3 +352,247 @@ def unscale(self, r):
+            inverse_mass_matrix[site_names] = _arrowhead_sqrt_inverse_to_inverse(sqrt_inverse)
+        return inverse_mass_matrix
+
+    @property
+    def mass_matrix(self):
+        return self._mass_matrix
+
+    @mass_matrix.setter
+    def mass_matrix(self, value):
+        for site_names, mass_matrix in value.items():
+            self._adapt_scheme[site_names].reset()
+            mass_matrix_sqrt = _arrowhead_sqrt(mass_matrix)
+            mass_matrix_sqrt_inverse = _arrowhead_sqrt_to_sqrt_inverse(mass_matrix_sqrt)
+            self._mass_matrix[site_names] = mass_matrix
+            self._mass_matrix_sqrt[site_names] = mass_matrix_sqrt
+            self._mass_matrix_sqrt_inverse[site_names] = mass_matrix_sqrt_inverse
+
+    def configure(self, mass_matrix_shape, adapt_mass_matrix=True, options={}):
+        """
+        Sets up an initial mass matrix.
+
+        :param dict mass_matrix_shape: a dict that maps tuples of site names to the shape of
+            the corresponding mass matrix. Each tuple of site names corresponds to a block.
+        :param bool adapt_mass_matrix: a flag to decide whether an adaptation scheme will be used.
+        :param dict options: tensor options to construct the initial mass matrix.
+        """
+        mass_matrix = {}
+        for site_names, shape in mass_matrix_shape.items():
+            size = shape[0]
+            self._mass_matrix_size[site_names] = size
+            # we set head_size=0 if diagonal, otherwise min(default_head_size, mass_matrix_size)
+            head_size = self._head_size if isinstance(self._head_size, int) else self._head_size[site_names]
+            head_size = 0 if len(shape) == 1 else min(head_size, size)
+            top = torch.eye(head_size, size, **options) * self._init_scale
+            bottom_diag = torch.full((size - head_size,), self._init_scale, **options)
+            mass_matrix[site_names] = ArrowheadMatrix(top, bottom_diag)
+            if adapt_mass_matrix:
+                adapt_scheme = WelfordArrowheadCovariance(head_size=head_size)
+                self._adapt_scheme[site_names] = adapt_scheme
+
+        self.mass_matrix = mass_matrix
+
+    def update(self, z, z_grad):
+        """
+        Updates the adaptation scheme using the new sample `z` or its grad `z_grad`.
+
+        :param dict z: the current value.
+        :param dict z_grad: grad of the current value.
+        """
+        for site_names, adapt_scheme in self._adapt_scheme.items():
+            z_grad_flat = torch.cat([z_grad[name].reshape(-1) for name in site_names])
+            adapt_scheme.update(z_grad_flat)
+
+    def end_adaptation(self):
+        """
+        Updates the current mass matrix using the adaptation scheme.
+        """
+        mass_matrix = {}
+        for site_names, adapt_scheme in self._adapt_scheme.items():
+            top, bottom_diag = adapt_scheme.get_covariance(regularize=True)
+            mass_matrix[site_names] = ArrowheadMatrix(top, bottom_diag)
```

Comment:
Do I understand correctly that 'covariance' here means "covariance of z gradients" rather than "covariance of samples"?